core/store/src/trie/mem/flexible_data/encoding.rs

```diff
@@ -0,0 +1,132 @@
+    /// encoder position, and then advance the position by the size of the
+    /// flexibly-sized part, as returned by `header.flexible_data_length()`.
+    pub unsafe fn encode_flexible<T: FlexibleDataHeader>(
+        &mut self,
+        header: &T,
+        data: T::InputData,
+    ) {
+        let length = header.flexible_data_length();
+        assert!(self.pos + length <= self.data.len());
+        header.encode_flexible_data(data, self.ptr());
+        self.pos += length;
+    }
+
+    /// Finishes the encoding process and returns a pointer to the allocated
+    /// memory. The caller is responsible for freeing the pointer later.
+    pub unsafe fn finish(self) -> *const u8 {
+        assert_eq!(self.pos, self.data.len());
+        Box::into_raw(self.data) as *const u8
+    }
+}
+
+/// Facilitates the decoding of flexibly-sized data.
+/// The lifetime 'a should be a lifetime that guarantees the availability of
+/// the memory being decoded from.
+pub struct RawDecoder<'a> {
+    data: *const u8,
+    pos: usize,
+    _marker: std::marker::PhantomData<&'a ()>,
+}
+
+impl<'a> RawDecoder<'a> {
+    /// Starts decoding from the given memory location. The location should be
+    /// a pointer returned by RawEncoder::finish.
+    pub fn new(data: *const u8) -> RawDecoder<'a> {
+        RawDecoder { data, _marker: std::marker::PhantomData, pos: 0 }
+    }
+
+    unsafe fn ptr(&mut self) -> *const u8 {
+        self.data.offset(self.pos as isize)
+    }
+
+    /// Decodes a fixed-size field at the current decoder position, and then
+    /// advances the position by the size of the field.
+    pub unsafe fn decode<T>(&mut self) -> &'a T {
+        let result = &*(self.ptr() as *const T);
+        self.pos += std::mem::size_of::<T>();
+        result
+    }
+
+    /// Like decode, but returns the reference as mutable.
+    pub unsafe fn decode_as_mut<T>(&mut self) -> &'a mut T {
+        let result = &mut *(self.ptr() as *mut T);
+        self.pos += std::mem::size_of::<T>();
+        result
+    }
+
+    /// Decodes a fixed-sized field at the current position, but does not
+    /// advance the position.
+    pub unsafe fn peek<T>(&mut self) -> &'a T {
+        &*(self.ptr() as *const T)
+    }
+
+    /// Decodes a flexibly-sized part of the data at the current position,
+    /// and then advances the position by the size of the flexibly-sized part,
+    /// as returned by `header.flexible_data_length()`.
+    pub unsafe fn decode_flexible<T: FlexibleDataHeader>(&mut self, header: &'a T) -> T::View<'a> {
+        let length = header.flexible_data_length();
+        let view = header.decode_flexible_data(self.ptr());
+        self.pos += length;
+        view
+    }
+
+    /// Takes a fixed field from the current position, and then advances the
+    /// position by the size of the field. This is useful for dropping the fixed
+    /// field when the overall structure is being dropped.
+    pub unsafe fn take_fixed<T>(&mut self) -> T {
```

Comment:
See above. Also, why not just use `std::ptr::read(self.ptr() as *mut T)`? I don’t see why you’d need to overwrite the current memory location with new uninitialized data; the old data is perfectly good uninitialized data :)